// @ggicci/goplay/dist/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const sleep = (n) => new Promise((res) => setTimeout(res, n));
export class GoPlayProxy {
    constructor(proxyUrl = '/goplay') {
        this.proxyUrl = proxyUrl;
    }
    raiseForStatus(resp) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!resp.ok) {
                const message = yield resp.text();
                throw new Error(message ? resp.statusText + ': ' + message : resp.statusText);
            }
        });
    }
    compile(sourceCode, goVersion) {
        return __awaiter(this, void 0, void 0, function* () {
            const form = new FormData();
            form.append('version', '2');
            form.append('withVet', 'true');
            form.append('body', sourceCode);
            const resp = yield fetch(`${this.proxyUrl}/_/compile?backend=${goVersion || ''}`, {
                method: 'POST',
                body: form,
            });
            yield this.raiseForStatus(resp);
            return yield resp.json();
        });
    }
    renderCompile(container, sourceCode, goVersion) {
        return __awaiter(this, void 0, void 0, function* () {
            container.replaceChildren(this.renderMessage('system', 'Waiting for remote server...'));
            const js = yield this.compile(sourceCode, goVersion);
            // Clear output.
            container.replaceChildren();
            // Render build error.
            if (js.Errors != '') {
                container.appendChild(this.renderMessage('error', js.Errors));
                container.appendChild(this.renderMessage('system', '\nGo build failed.'));
                return;
            }
            // Render events.
            for (const event of js.Events || []) {
                container.appendChild(yield this.renderEvent(event));
            }
            container.appendChild(this.renderMessage('system', '\nProgram exited.'));
        });
    }
    renderEvent(e) {
        return __awaiter(this, void 0, void 0, function* () {
            if (e.Delay >= 0) {
                yield sleep(e.Delay / 1000000);
            }
            return this.renderMessage(e.Kind, e.Message);
        });
    }
    renderMessage(kind, message) {
        const output = document.createElement('span');
        output.classList.add(kind);
        output.innerText = message;
        return output;
    }
    share(sourceCode, goVersion) {
        return __awaiter(this, void 0, void 0, function* () {
            const resp = yield fetch(`${this.proxyUrl}/_/share`, {
                method: 'POST',
                body: sourceCode,
            });
            yield this.raiseForStatus(resp);
            const canonical = 'https://go.dev/play/p/' + (yield resp.text());
            return goVersion ? `${canonical}?v=${goVersion}` : canonical;
        });
    }
}
